index.js

```diff
@@ -53,14 +62,101 @@ var initConfig = function(projectConfig) {
     }
   };
   var config = extend(true, {}, require('./config.js'), projectConfig, envConfig);
+  // Enforce the upper limit for the label value size.
   if (config.maximumLabelValueSize > constants.TRACE_SERVICE_LABEL_VALUE_LIMIT) {
     config.maximumLabelValueSize = constants.TRACE_SERVICE_LABEL_VALUE_LIMIT;
   }
+  // Clamp the logger level.
+  if (config.logLevel < 0) {
+    config.logLevel = 0;
+  } else if (config.logLevel >= common.logger.LEVELS.length) {
+    config.logLevel = common.logger.LEVELS.length - 1;
+  }
   return config;
-};
+}
 
-var traceApi = new TraceApi('Custom Span API');
-var agent;
+/**
+ * Retrieves various fields from the GCP metadata service, and calls
+ * setMetadata on the TraceWriter singleton with these values.
+ * @param {common.logger} logger A logger object.
+ * @param {Boolean} getProjectId Whether the project ID should be retrieved
+ * from the metadata service.
+ */
+function getMetadata(logger, getProjectId) {
+  // Headers for GCP metadata requests
+  var headers = {};
+  headers[constants.TRACE_AGENT_REQUEST_HEADER] = 1;
+
+  var metadata = {};
+  var getHostnameAndInstanceId = function() {
+    gcpMetadata.instance({
+      property: 'hostname',
+      headers: headers
+    }, function(err, response, hostname) {
+      if (err && err.code !== 'ENOTFOUND') {
```

Comment:
You didn't modify this, but I find it disturbing that we lack symmetry in how we deal with `err` between here and how get the projectId.

Given that these pieces of code are together now, it would be good to make error handling consistent. If you don't want to make these changes as part of this PR, please open an issue for this.